src/coreclr/tools/Common/TypeSystem/Interop/MarshalAsDescriptor.cs

```diff
@@ -48,17 +48,24 @@ public enum NativeTypeKind : byte
 
     public class MarshalAsDescriptor
     {
+        private TypeDesc customMarshallerType;
+        private string cookie;
+
         public NativeTypeKind Type { get; }
         public NativeTypeKind ArraySubType { get; }
         public uint? SizeParamIndex { get; }
         public uint? SizeConst { get; }
+        public TypeDesc CustomMarshallerType => Type == NativeTypeKind.CustomMarshaler ? customMarshallerType : throw new InvalidOperationException("Custom marshaller type can be set only when using for CustomMarshaller");
```

Comment:
We don't do throws like this because then it just raises questions whether the message needs to be localized, etc. This should be an assert, but I'm not firm on the validation (we don't validate the array stuff above either).